### PR TITLE
[21206] Add basic checkout instructions

### PR DESCRIPTION
**Pull request by oliverguenther**
This commit adds very basic checkout instructions similar
to `openproject-checkout` to repositories.
### Issues
- Uses table style for checkout component, as exact styling is unclear
  - Adds compressed table style with less padding
- Checkout settings are best organized in a Hash with keys per SCM vendor. Using default settings, a lot of new setting values would be introduced. While this works fine, there are some caveats:
  - Checks for Hash in SettingsController
  - Currently can't use SettingsHelper, as it requires single setting
    values
### To decide
- [ ]  subtree checkout (May be out of scope, even though service functionality is there)
#### Checkout instructions

bildschirmfoto 2015-08-20 um 18 22 48
#### Checkout instructions settings

bildschirmfoto 2015-08-20 um 18 32 49

Relevant work package: https://community.openproject.org/work_packages/21206

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-21206-integrate-checkout.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-21206-integrate-checkout.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: lib/open_project/scm/manageable_repository.rb [hunk 1]
```diff
@@ -100,8 +100,6 @@ def managed_repository_url
         "file://#{managed_repository_path}"
       end
 
-      protected
```
Comment (oliverguenther) [inline lib/open_project/scm/manageable_repository.rb:103]: Removed in order to be able to use `project_identifier` in service class

**Comment by NobodysNightmare** [comment]
If I see this correctly the current implementation stores the informational text in a `Setting`. This has the caveat, that you won't be able to translate this text into the users locale.

The very generic default text for a vendor seems to be a prime candiate for a localization (_Please consult the documentation of %{vendor} if you need..._). The acceptance criteria only talk about instructions that "are defined by the SCM vendor implementations themselves", so is it really neccessary/desirable to be able to configure the whole instruction text?

I see that I want to be able to write something like "login using your GitHub credentials", but then again I would wish for localizability.

**Comment by oliverguenther** [comment]
We could probably add the default text using I18n key. Users wishing to override that text would specify that as `user_text`. In the implementation, you check for `user_text` or fallback to `I18n.t(default_text)`. 

I'm not sure If we need to override the text in our installations, but its rather simple behavior that I wouldn't simply replace with a fixed translation.

My proposed change would also allow users to delete their custom text again, and it falls back to the default (currently, the placeholder text is lost upon once deleted)-

**Comment by NobodysNightmare** [comment]
That sounds like a reasonable middle ground...

File: app/services/scm/checkout_instructions_service.rb [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  end
+
+  def checkout_enabled?
+    checkout_settings['enabled'].to_i > 0
+  end
+
+  ##
+  # Determines whether permissions for the given repository
+  # are available.
+  def permission?
```
Comment (NobodysNightmare) [inline app/services/scm/checkout_instructions_service.rb]: At first this reads like the worst permission check ever, like

``` ruby
if permission?
  # I have the permission
else
  # I don't have it :(
end
```

I know I tend to use longer names, but `manges_permissions?`?

I'd also use the plural form for both permission relevant methods: `permissions`

File: app/services/scm/checkout_instructions_service.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  end
+
+  ##
+  # Returns whether the given user may commit to the repository
+  #
+  # Note that this information is only applicable when the repository is managed,
+  # because otherwise OpenProject does not control the repository permissions.
+  # Use +permission?+ to check whether this is the case.
+  def may_commit?
+    permission == :readwrite
```
Comment (NobodysNightmare) [inline app/services/scm/checkout_instructions_service.rb:132]: I somehow see `permission` returning an array of either `[]`, `[:read]` or `[:read, :write]`... I think you are not doing it, because the way you are doing it now is probably better suited to generate a human readable string...

I am not sure which way I would prefer...

edit: Both permission checks would then basically look the same, like: `permission.include? :write`

File: app/views/repositories/_checkout_instructions.html.erb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            <div class="form--text-field-container -xwide">
+                <input type="text" class="form--text-field"
+                onclick="this.focus(); this.select()"
+                readonly="readonly" value="<%= instructions.checkout_url %>">
+            </div>
+        </div>
+        <div class="form--field-instructions -no-italic">
+          <%= simple_format instructions.instructions %>
+          <% if instructions.permission? %>
+            <p>
```
Comment (NobodysNightmare) [inline app/views/repositories/_checkout_instructions.html.erb:21]: [close me :(](https://xkcd.com/859/)

File: app/views/repositories/empty.html.erb [hunk 5]
```diff
@@ -33,6 +33,7 @@ See doc/COPYRIGHT.rdoc for more details.
              message: l('repositories.errors.empty_repository')) %></p>
   </div>
 </div>
+<%= render partial: 'checkout_instructions', locals: { repository: @repository, instructions: @instructions } %>
```
Comment (NobodysNightmare) [inline app/views/repositories/empty.html.erb:36]: This is nitpickery, but I'd expect the instructions above the empty message. Probably because I am used to empty tables being displayed as

> some stuff above
> _no data to display_

edit: Finally found the example I was looking for

image

File: app/views/settings/_repositories_checkout.html.erb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+<fieldset class="form--fieldset">
+  <legend class="form--fieldset-legend"><%= l('repositories.checkout.instructions') %></legend>
+  <% OpenProject::Scm::Manager.registered.each do |vendor, klass| %>
+      <% vendor = vendor.to_s %>
+      <% setting_base = "settings[repository_checkout_data][#{vendor}]" %>
+      <% enabled = Setting.repository_checkout_data[vendor]['enabled'].to_i > 0 %>
+      <fieldset class="form--fieldset -collapsible collapsed">
+        <legend class="form--fieldset-legend" title="<%= klass.vendor_name %>" onclick="toggleFieldset(this);">
+          <a href="javascript:"><%= klass.vendor_name %>
+            <span class="fieldset-toggle-state-label hidden-for-sighted">expanded</span>
```
Comment (NobodysNightmare) [inline app/views/settings/_repositories_checkout.html.erb]: I18n: expanded

File: app/assets/stylesheets/content/_forms.sass [hunk 7]
```diff
@@ -328,6 +331,9 @@ fieldset.form--fieldset
   @extend %form--field-after-container
   font-style: italic
 
+  &.-no-italic
+    font-style: normal
```
Comment (NobodysNightmare) [inline app/assets/stylesheets/content/_forms.sass:335]: One issue I have with this modifier is that its name carries no semantic information (as in: why should I use no italics), but only the mere styling information: "use no italics, don't ask why"...

Still I have no better idea :/

File: config/locales/en.yml [hunk 8]
```diff
@@ -1286,7 +1286,18 @@ en:
...
     autofetch_information: "Check this if you want repositories to be updated automatically when accessing the repository module page.\nThis encompasses the retrieval of commits from the repository and refreshing the required disk storage."
-    checkout_instructions: "Checkout instructions"
+    checkout:
+      access:
+        readwrite: 'Read + Write'
+        read: 'Read-only'
+        none: 'No checkout access, you may only view the repository through this application.'
+      access_permission: 'Your permissions on this repository:'
+      url: "Checkout URL"
+      base_url_text: "The base URL to use for generating checkout URLs (e.g., https://myserver.example.org/repos/).\nNote: The base URL is only used for rewriting checkout URLs in managed repositories. Other repositories are not altered"
```
Comment (NobodysNightmare) [inline config/locales/en.yml]: I think this string should end in punctuation, as it also has intermediate punctuation.

File: app/services/scm/checkout_instructions_service.rb (hunk 3 above)
Comment (oliverguenther) [inline app/services/scm/checkout_instructions_service.rb:132]: I'm not yet convinced of what the advantage would be. If the permissions would be independent from another, then I agree that it might make sense to return `[read,write]`. 

In my case however, `:readwrite` is a superset of permissions of `:read` and returning a single symbol makes things a lot simpler outside the service, as it maintains the checks inside the service.

File: app/views/repositories/empty.html.erb (hunk 5 above)
Comment (oliverguenther) [inline app/views/repositories/empty.html.erb:36]: I'm open for discussions here, but I actually prefer it otherwise, as I parse the page top-down, and should receive the warning first.

bildschirmfoto 2015-08-24 um 16 56 58

bildschirmfoto 2015-08-24 um 16 57 11

File: app/views/settings/_repositories_checkout.html.erb (hunk 6 above)
Comment (oliverguenther) [inline app/views/settings/_repositories_checkout.html.erb]: Copy/paste error, that toggler is added dynamically.

File: app/views/repositories/empty.html.erb (hunk 5 above)
Comment (NobodysNightmare) [inline app/views/repositories/empty.html.erb:36]: You are right. Since we are displaying it as a warning I agree.

We could discuss whether the warning is apropriate here, though. Since everything is fine, except for the repository being empty, which is the default for the repository of a new project.

We do not say

> :warning: There are no work packages yet.

do we?

Comment (oliverguenther) [inline app/views/repositories/empty.html.erb:36]: I agree, we have a common pattern (e..g, in work packages) to display this larger notification / block with the empty 'warning'. We could apply this here, too, and then I'd agree to move the checkout above it.

However this is rather out of scope here.

**Comment by NobodysNightmare** [comment]
When hovering over the text field with the checkout URL my cursor turns into a little forbidden symbol forbidden.

I find this quite irritating, since I feel the urge not to click into that field anymore... it is not allowed after all...

**Comment by oliverguenther** [comment]
I think this is the default style of 'readonly', but I agree.

File: app/services/scm/checkout_instructions_service.rb (hunk 2 above)
Comment (oliverguenther) [inline app/services/scm/checkout_instructions_service.rb]: `manages_permission?` :+1:

**Comment by NobodysNightmare** [comment]
There is a failure on travis:

```
  1) OpenProject::Storage with no SCM configuration #mount_information should contain one fs entry
     Failure/Error: expect(subject.length).to be == 1
       expected: == 1
            got:    0
     # ./spec/lib/open_project/storage_spec.rb:55:in `block (4 levels) in <top (required)>'
```

Besides that the PR LGTM.

**Comment by oliverguenther** [comment]
Really seemed to be unrelated to this PR. I have an idea concerning the flickering storage spec, addressing this here https://github.com/opf/openproject/pull/3410

**Comment by NobodysNightmare** [comment]
> - [ ]  subtree checkout (May be out of scope, even though service functionality is there)

I'd say even if we want to do this until 5.0, we can safely do it in a separate PR.

I am therefore merging this PR now.
